general/mem_manager.hpp

```diff
@@ -175,14 +175,28 @@ class Memory
    /// Copy constructor: default.
    Memory(const Memory &orig) = default;
 
-   /// Move constructor: default.
-   Memory(Memory &&orig) = default;
+   /// Move constructor, takes ownership when possible and invalidates @a orig.
+   Memory(Memory &&orig)
+      : h_ptr(orig.h_ptr), capacity(orig.capacity), h_mt(orig.h_mt),
+        flags(orig.flags)
+   {
+      orig.Reset();
+   }
```

Comment:
```suggestion
   Memory(Memory &&orig)
   {
      *this = orig;
      orig.Reset();
   }
```

Reply:
What I don't like with this implementation is that it relies on the copy assignment operator to create "multiple" owners which I don't like and would like to change in the future since it is bug prone (double `free`, usage of data after `free`, etc...).
I'm also wondering if we should call `Delete`, since it's potentially a memory leak not to, and I can't think of a good reason not to call `Delete`.

Reply:
Both versions are functionally the same (there are two owners immediately before `orig.Reset()` in both versions of the code). The advantage of what I suggested is that it should work without modification even if you add more data members.

It is not the role of the move constructor to destroy the object. Its destructor will be called eventually (usually very soon).

Reply:
What I meant is that the version you're proposing will become wrong if we change the behavior of the copy. But I'm okay using the copy to write the move if it's important to you.

I don't think you can discard calling `Delete()` that easily, you have to remember that the destructor does not call `Delete()`, so using move semantics without calling first `Delete()` would result in a memory leak, and I do not see any benefits in letting that ever happen.